### PR TITLE
Adding unit tests for IndexedDataVector

### PR DESCRIPTION
**Pull request by invertednormal**
Also moving the tests for MultiIndexedDataVector currently in IndexedDataVectorTests to MultiIndexedDataVectorTests.

**Reviewer Note - File Rename**: The code in IndexedDataVectorTests is all new, don't bother looking at the comparison. The code that used to be in IndexedDataVectorTests has been moved to MultiIndexedDataVectorTests and only has minimal updates. This was done as two separate commits, so its easier to look at those individual diffs instead of the diff for the whole PR. (See d4a59a388491be7dbfd6c558704894a5377c6c81 and 4ae837410127b5428d3de8cf1ca31ceeaeee2629)

### CODE REVIEW

**Review by mbalfour-amzn — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Tests/IndexedDataVectorTests.cpp [hunk 1]
```diff
@@ -8,338 +8,218 @@
...
+                    data.GetData(index) = value;
+                    value += incrementAmount;
+                }
+            }
+            return data;
         }
 
-        void DestroyAllocator()
+        template<typename T>
+        void ShuffleIndexedDataVector(IndexedDataVector<T> dataVector, AZStd::vector<uint16_t> indices)
```
Comment (santorac) [inline Gems/Atom/Feature/Common/Code/Tests/IndexedDataVectorTests.cpp:57]: I think you meant to pass dataVector by reference. It's currently passed by value so this function doesn't actually change anything.

**Review by santorac — CHANGES_REQUESTED** [review]

**Comment by santorac** [comment]
I know you are just adding a unit test here, but the class IndexedDataVector really needs some API documentation. There is no explanation for what each of the functions does.

**Comment by santorac** [comment]
I'm also finding that the code in IndexedDataVector is difficult to read because of the overloaded use of the term "index". It would be easier if we used terms like "logical index" and "physical index".

File: Gems/Atom/Feature/Common/Code/Tests/IndexedDataVectorTests.cpp (hunk 1 above)
Comment (invertednormal) [reply Gems/Atom/Feature/Common/Code/Tests/IndexedDataVectorTests.cpp]: Good catch. Thankfully the unit tests still worked :)

**Comment by invertednormal** [comment]
> I know you are just adding a unit test here, but the class IndexedDataVector really needs some API documentation. There is no explanation for what each of the functions does. (Unless comments were added on the development branch. I'm looking at stabilization).

Fair point. I went ahead and added some documentation with this PR, easy enough to take care of now.

**Comment by invertednormal** [comment]
> I'm also finding that the code in IndexedDataVector is difficult to read because of the overloaded use of the term "index". It would be easier if we used terms like "logical index" and "physical index".

Holding off on changes here for now since that will have larger knock-on effects. IMO the only confusing call is GetRawIndex() since it's the only function that refers to an "index" that's the physical index, not the logical index. Maybe we could just rename this function to something like "GetPhysicalIndex()" or "GetDataOffset()" but keep the rest the same?

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Utils/IndexedDataVector.h [hunk 2]
```diff
@@ -27,31 +27,49 @@ namespace AZ::Render
...
 
     private:
         constexpr static size_t InitialReservedSize = 128;
 
-        // Stores data indices and an embedded free list
+        // Indices to data and an embedded free list in the unused entries
         AZStd::vector<IndexType> m_indices;
-        // Stores the indirection index
+
+        // Map of index in m_data to the index for that data in m_indices.
```
Comment (santorac) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Utils/IndexedDataVector.h:69]: This is a good example of where the "physical" and "logic" distinction would be helpful.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Utils/IndexedDataVector.h [hunk 3]
```diff
@@ -27,31 +27,49 @@ namespace AZ::Render
...
+        //! Returns a reference to the internal vector.
+        const AZStd::vector<IndexType>& GetDataToIndexVector() const;
 
-        AZStd::vector<IndexType>& GetIndexVector();
-        const AZStd::vector<IndexType>& GetIndexVector() const;
-
+        //! Returns the offset into the internal data vector for a given index.
         IndexType GetRawIndex(IndexType index) const;
+
+        //! Returns the public index for data given its pointer.
```
Comment (santorac) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/Utils/IndexedDataVector.h:60]: The part I found confusing at first was that m_indices values might be physical indexes when the data is present, or might be logical indexes into itself to form the linked list of free spots.
I think the comment for that member could still be expanded a bit

**Review by santorac — APPROVED** [review]

**Pull request merged**